aslam_cv_cameras/src/camera-3d-lidar.cc

```diff
@@ -0,0 +1,185 @@
+  kp[1] = verticalCenter() - (kp[1] * verticalResolution());
+  (*out_point_3d)[0] = -std::sin(kp[0]) * std::cos(kp[1]);
+  (*out_point_3d)[1] = -std::sin(kp[1]);
+  (*out_point_3d)[2] = std::cos(kp[0]) * std::cos(kp[1]);
+  out_point_3d->normalize();
+  return true;
+}
+
+const ProjectionResult Camera3DLidar::project3Functional(
+    const Eigen::Ref<const Eigen::Vector3d>& point_3d,
+    const Eigen::VectorXd* intrinsics_external,
+    const Eigen::VectorXd* distortion_coefficients_external,
+    Eigen::Vector2d* out_keypoint,
+    Eigen::Matrix<double, 2, 3>* out_jacobian_point,
+    Eigen::Matrix<double, 2, Eigen::Dynamic>* out_jacobian_intrinsics,
+    Eigen::Matrix<double, 2, Eigen::Dynamic>* out_jacobian_distortion) const {
+  CHECK_NOTNULL(out_keypoint);
+
+  CHECK(distortion_coefficients_external == nullptr)
+      << "External distortion is not implemented!";
+  CHECK(out_jacobian_point == nullptr)
+      << "Jacobians are currently not implemented!";
+  CHECK(out_jacobian_intrinsics == nullptr)
+      << "Jacobians are currently not implemented!";
+  CHECK(out_jacobian_distortion == nullptr)
+      << "Jacobians are currently not implemented!";
+
+  // Determine the parameter source. (if nullptr, use internal)
+  const Eigen::VectorXd* intrinsics;
+  if (!intrinsics_external)
+    intrinsics = &getParameters();
+  else
+    intrinsics = intrinsics_external;
+  CHECK_EQ(intrinsics->size(), kNumOfParams) << "intrinsics: invalid size!";
+
+  // This camera does not have distortion, so we do not use this.
+  const Eigen::VectorXd dummy_distortion_coefficients;
+
+  return project3Functional<double, NullDistortion>(
+      point_3d, *intrinsics, dummy_distortion_coefficients, out_keypoint);
+}
+
+Eigen::Vector2d Camera3DLidar::createRandomKeypoint() const {
+  Eigen::Vector2d out;
+  out.setRandom();
+  // Unit tests often fail when the point is near the border. Keep the point
+  // away from the border.
+  double border = std::min(imageWidth(), imageHeight()) * 0.1;
+
+  out(0) = border + std::abs(out(0)) * (imageWidth() - border * 2.0);
+  out(1) = border + std::abs(out(1)) * (imageHeight() - border * 2.0);
+
+  return out;
+}
+
+Eigen::Vector3d Camera3DLidar::createRandomVisiblePoint(double depth) const {
+  CHECK_GT(depth, 0.0) << "Depth needs to be positive!";
+  Eigen::Vector3d point_3d;
+
+  Eigen::Vector2d y = createRandomKeypoint();
+  backProject3(y, &point_3d);
+  point_3d /= point_3d.norm();
+
+  // Muck with the depth. This doesn't change the pointing direction.
+  return point_3d * depth;
+}
+
+void Camera3DLidar::getBorderRays(Eigen::MatrixXd& rays) const {
+  rays.resize(4, 8);
+  Eigen::Vector4d ray;
+  backProject4(Eigen::Vector2d(0.0, 0.0), &ray);
+  rays.col(0) = ray;
+  backProject4(Eigen::Vector2d(0.0, imageHeight() * 0.5), &ray);
+  rays.col(1) = ray;
+  backProject4(Eigen::Vector2d(0.0, imageHeight() - 1.0), &ray);
+  rays.col(2) = ray;
+  backProject4(Eigen::Vector2d(imageWidth() - 1.0, 0.0), &ray);
```

Comment:
yes, but this is not a performance issue since the compiler will inline them anyways. But I guess it's a style/interface decision, since this member is inherited from the parent it tries to use them through the getters. I just followed the aslam_cv style, it's done like this for all camera parameters in all functions.